### PR TITLE
Split into library - fixes #2

### LINKED ISSUE #2: Split out the core logic as a library
This is needed so the bitbucket-pipline variant can be build on top of this

### PR DESCRIPTION
**Pull request by EugenMayer**
Split project into 'core' and 'cli' modules - fixed #2 

- add github actions pipeline
- automate tests

### CODE REVIEW

**Comment by markrekveld** [comment]
I pushed a tab character fix to main, a simple rebase should get this branch back in order

**Comment by markrekveld** [comment]
@EugenMayer the main test classes `BaseTestPublishToMarketplace.java` , `CommandLinePublishToMarketplaceTest`, `EnvironmentVariablesPublishToMarketplaceTest` are not in a java source directory so are not executed.
Once they are in a test source directory, they will also need to have classpath access the fake-jira and fake-marketplace test resources for the core package. This can be done by adding the core tests artifact as test dependency on the cli module.

File: Makefile
Comment (markrekveld) [inline Makefile:1]: For my education; why use a make file?

File: Makefile
Comment (markrekveld) [inline Makefile]: Would it not be easier to just build the whole project instead of the parent followed by all modules?

Comment (markrekveld) [reply Makefile]: Also relates to the use of a make file I guess

File: pom.xml
Comment (markrekveld) [inline pom.xml]: For my education; Do you have a specific reason for moving the scm element to the end of the file.

File: Makefile
Comment (EugenMayer) [reply Makefile:1]: Well mainly since it is an easy interface for

1. chaining tasks like `make build test` and define task dependencies like `build: build-parent build-core` without installing any frameworks, tooling or anything
2. Using some tool of this sort lets me bundle all build task in one place and reused them for devs and ci, which reduces frictions, keeps it DRY and reproducable

mvn itself is kind of hard to use for those orchestrations and using bash would lose point 1. (still give 2.)

File: Makefile
Comment (EugenMayer) [reply Makefile]: how would you build the whole project? I find myselfs in hot water with maven and composite projects over and over again. Means

- i build the project using `mvn --also-make -pl cli clean package` (first build, all fine)
- then i do changes on core
- run `mvn --also-make -pl cli clean package`

And what usually happens is that mvn will not rebuild core OR the build of cli will not include the last artifact. Those issues with mvn and composite is one of the main reasons i use gradle, because this is entirely handled there (you consume composite builds using projects artifacts ... not pseudo coordinates with a magic version number that matches some local `0-SNAPSHOT`.

So IMHO, i use a clean rebuild of each module every time in an manual, orchestrated way, so that mvn will not screw up. This ensure i always test what i intend to.

If you have a better idea or a tip, a magic flag to fix this, please share it with me - but that above is basically my go-to with mvn projects as "better safe then outdated"

File: pom.xml
Comment (EugenMayer) [reply pom.xml]: No not at all, i assume it was some kind of preference like "the things you need for secondary maintenance like dep-management/properties as far top up, things you do not change regularly like repos/scm' far down. There was no other rational here - if it was too ballsy, let me know, i try to place it back were it was.

**Comment by EugenMayer** [comment]
@markrekveld not sure why you pushed a code-formatting commit on main before merging all branches, but the above merge conflict is yours to take please :)

**Comment by markrekveld** [comment]
> not sure why you pushed a code-formatting commit on main before merging all branches

So that your P doesn't include reformatting, making it easier to review.
I pushed a commit with the rebase

**Comment by EugenMayer** [comment]
we will need to rebase the other PR too, i can try to do that. Beside that, i was not going to change code here besides imports if any possible, so no formatting. Is there any way to share your codestyles so we can be sure we work with the same?

**Comment by markrekveld** [comment]
@EugenMayer I have [this repo](https://bitbucket.org/marvelution/codestyle) where I store my code style profiles. All are exported from my IntelliJ IDEA

**Comment by markrekveld** [comment]
> we will need to rebase the other PR too, i can try to do that.

True, you can do it if you like, or leave it to me since I'll be updating the PR anyway to support my maven repositories

File: Makefile
Comment (markrekveld) [reply Makefile]: > how would you build the whole project? 

Basically you call `mvn clean package` and it will build the main project and all modules. You would use a flag like `-pl` to only build a specific module

> And what usually happens is that mvn will not rebuild core OR the build of cli will not include the last artifact.

If you do a `clean package` then the artifact is always updated, if you just do `package` then it can happen that artifact is not completely up to date, all though I have only seen this happen when I save file mid build.

> If you have a better idea or a tip, a magic flag to fix this, please share it with me

On CI systems I always do a `clean install` make sure no old build files are used and that the actifacts are installed on the local m2 repo for later use.
Locally it depends on what I need, just a compile to debug some tests I just call `compile` otherwise its mostly either `install` or `clean install`

Comment (EugenMayer) [reply Makefile]: thanks for your insight. i adjusted the task 

```
build-all:
	./mvnw $(MVN_FLAGS) -DskipTests clean install
```

so when building the entire project, it will use you recommendation. There is still build-parent, build-core,build-cli for specific builds, but the CI and the readme generally use `make build-all`

Does this cover your suggestion or would there be more?

**Comment by EugenMayer** [comment]
@markrekveld i would love to finish this one up - are any things open for you or could we potentially plan a merge?

File: Makefile
Comment (markrekveld) [reply Makefile]: Yeah that covers it

**Comment by markrekveld** [comment]
I updated the GH action to runon worklfow_dispatch instead of push since I have not yet had the chance to update the CI/CD infra to support GH

**Review by markrekveld — DISMISSED** [review]

**Comment by EugenMayer** [comment]
do you mean your nexus by that? if yes, i would rather stop deploying snapshots but let the gh action run, so we have tests and builds. What do you think?

**Comment by markrekveld** [comment]
> do you mean your nexus by that? if yes, i would rather stop deploying snapshots but let the gh action run, so we have tests and builds. What do you think?

Yeah, I need to provision temporal permissions for GH and don't have the time right now, its a bit more then just enabling.
If the build can run without issue, then sure we can enable it for push triggers.

**Comment by EugenMayer** [comment]
This PR just builds, it does not deploy - so for this PR you do not need to disable it at all. Could you just enable it again and we might just wait mergin the other pr (release/publihs) pr until you find time to provision the secerts

**Comment by markrekveld** [comment]
> This PR just builds, it does not deploy - so for this PR you do not need to disable it at all. Could you just enable it again and we might just wait mergin the other pr (release/publihs) pr until you find time to provision the secerts

I just enabled push builds again.

**Comment by EugenMayer** [comment]
Thanks. So are we good top go then? (just give your blessing via a review as soon as you have time and i merge it)

**Review by markrekveld — APPROVED** [review]

**Comment by markrekveld** [comment]
I needed to update the review ruleset and strait afterwards merged it, sorry I read over the 'i can merge' part of your last comment.

**Comment by EugenMayer** [comment]
no worries, i do not mind at all. Work is done - me happy!

**Pull request merged**
